libraries/CryptoInterface/src/CryptoInterface.h

```diff
@@ -0,0 +1,806 @@
+     * 
+     * The output values of chacha20Poly1305Encrypt should be passed as input values to chacha20Poly1305Decrypt.
+     * 
+     * Note that a 12 byte nonce is generated via getNonceGenerator() every time chacha20Poly1305Encrypt is called.
+     * If the same key and nonce combination is used more than once for distinct messages, the encryption will be broken, so keep the following in mind:
+     * 
+     * By default the nonce is generated via the hardware random number generator of the ESP8266. 
+     * The entropy of this source may not be sufficient to avoid nonce collisions, so to further reduce the risk of encryption failure
+     * it is recommended that a keySalt is always provided when using the default nonceGenerator. Using a keySalt will create a 
+     * pseudorandom subkey from the original key via HKDF, and use that for the encryption/decryption.
+     * The same key + keySalt will always generate the same subkey.
+     * 
+     * An alternative to using a keySalt is to change the nonceGenerator so that it does not rely on random numbers.
+     * One way to do this would be to use a counter that guarantees the same key + nonce combination is never used.
+     * This may not be easily achievable in all scenarios, however.
+     * 
+     * @param data An array containing the data to encrypt. The encrypted data is generated in place, so when the function returns the data array will contain the encrypted data.
+     * @param dataLength The length of the data array in bytes.
+     * @param key The secret encryption key to use. Must be 32 bytes (ENCRYPTION_KEY_LENGTH) long.
+     * @param keySalt The salt to use when generating a subkey from key. Note that the same salt must be used during decryption as during encryption. Set to nullptr to prevent subkey generation.
+     * @param keySaltLength The length of keySalt in bytes.
+     * @param resultingNonce The array that will store the nonce generated during encryption. Must be able to contain at least 12 bytes. The nonce is not secret and must be passed to the decryption function.
+     * @param resultingTag The array that will store the message authentication tag generated during encryption. Must be able to contain at least 16 bytes. The tag is not secret and must be passed to the decryption function.
+     * @param aad Additional authenticated data. This data will be covered by the Poly1305 MAC, but not encrypted. 
+     *            You can include the unencrypted parts of your message as AAD to ensure that the encrypted content cannot 
+     *            be re-sent with replaced unencrypted data by an attacker.
+     *            Defaults to nullptr.
+     * @param aadLength The length of the aad array in bytes. Defaults to 0.
+     */
+    void chacha20Poly1305Encrypt(void *data, const size_t dataLength, const void *key, const void *keySalt, const size_t keySaltLength, void *resultingNonce, void *resultingTag, const void *aad = nullptr, const size_t aadLength = 0);
```

Comment:
Same as above.  ChaCha is fast on the 8266, but AES is also widely used and abstracting one more layer into an encryption major class seems more logical to me.

Reply:
There is now an object with encrypt and decrypt methods. Since free standing objects [were requested ](https://github.com/esp8266/Arduino/pull/6961#discussion_r415436174) no further abstraction has been done.